File: arm.64/frame/qt/src/bind.cpp
#include "type.h"
#include <filesystem>
#include <dlfcn.h>

namespace fs = std::filesystem;

Cabi* bind(const std::string& binary_path) {
    std::string filename = fs::path(binary_path).filename().string();
    if (filename == "c" || filename == "cpp" || filename == "rust" ||
        filename == "go" || filename == "swift" || filename == "haskell" ||
        filename == "zig" || filename == "v") {
        return nullptr;
    }

    void* lib = dlopen(binary_path.c_str(), RTLD_LAZY);
    if (!lib) {
        return nullptr;
    }

    auto more = reinterpret_cast<MoreFn>(dlsym(lib, "More"));
    auto pump = reinterpret_cast<PumpFn>(dlsym(lib, "Pump"));
    auto less = reinterpret_cast<LessFn>(dlsym(lib, "Less"));

    if (!more || !pump || !less) {
        dlclose(lib);
        return nullptr;
    }

    auto* cabi = new Cabi();
    cabi->name = filename;
    cabi->path = binary_path;
    cabi->lib  = lib;
    cabi->More = more;
    cabi->Pump = pump;
    cabi->Less = less;

    return cabi;
}

File: arm.64/frame/metal/src/bind.swift
import Foundation

public func bind(binaryPath: String) -> Cabi? {
    let filename = (binaryPath as NSString).lastPathComponent
    let skip = ["c", "cpp", "rust", "go", "swift", "haskell", "zig", "v"]
    if skip.contains(filename) {
        return nil
    }

    guard let lib = dlopen(binaryPath, RTLD_LAZY) else {
        return nil
    }

    guard let moreSym = dlsym(lib, "More"),
          let pumpSym = dlsym(lib, "Pump"),
          let lessSym = dlsym(lib, "Less") else {
        dlclose(lib)
        return nil
    }

    let more = unsafeBitCast(moreSym, to: MoreFn.self)
    let pump = unsafeBitCast(pumpSym, to: PumpFn.self)
    let less = unsafeBitCast(lessSym, to: LessFn.self)

    return Cabi(
        name: filename,
        path: binaryPath,
        lib: lib,
        More: more,
        Less: less,
        Pump: pump
    )
}
